### PR TITLE
refactor(compiler): rework and expose APIs to be used in schematics

### PR DESCRIPTION
**Pull request by crisbeto**
Reworks some of the existing compiler APIs to make them easier to use in a schematic and exposes a few new ones to surface information we already had. High-level list of changes:
* `getPotentialImportsFor` now requires a class reference, instead of a `PotentialDirective | PotentialPipe`.
* New `getNgModuleMetadata` method has been added to the type checker.
* New `getPipeMetadata` method has been added to the type checker.
* New `getUsedDirectives` method has been added to the type checker.
* New `getUsedPipes` method has been added to the type checker.
* The `decorator` property was exposed on the `TypeCheckableDirectiveMeta`. The property was already present at runtime, but it wasn't specified on the interface.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts [hunk 1]
```diff
@@ -732,7 +757,7 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
     const exportingNgModules = this.ngModuleIndex.getNgModulesExporting(meta.ref.node);
     if (exportingNgModules !== null) {
       for (const exporter of exportingNgModules) {
-        const emittedRef = this.emit(PotentialImportKind.Standalone, exporter, inContext);
+        const emittedRef = this.emit(PotentialImportKind.NgModule, exporter, inContext);
```
Comment (crisbeto) [inline packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:760]: @dylhunn this change is unrelated to what I need for the schematics, but it seemed like an oversight from the previous PR, unless I'm misunderstanding it.

File: packages/compiler/src/render3/view/t2_binder.ts [hunk 2]
```diff
@@ -331,8 +331,6 @@ class DirectiveBinder<DirectiveT extends DirectiveMeta> implements Visitor {
 class TemplateBinder extends RecursiveAstVisitor implements Visitor {
   private visitNode: (node: Node) => void;
 
-  private pipesUsed: string[] = [];
```
Comment (crisbeto) [inline packages/compiler/src/render3/view/t2_binder.ts:334]: Unrelated change. I just noticed that it's not being used.

File: packages/compiler-cli/src/ngtsc/typecheck/api/checker.ts [hunk 3]
```diff
@@ -150,8 +153,8 @@ export interface TemplateTypeChecker {
    * In the context of an Angular trait, generate potential imports for a directive.
    */
   getPotentialImportsFor(
-      toImport: PotentialDirective|PotentialPipe,
-      inComponent: ts.ClassDeclaration): ReadonlyArray<PotentialImport>;
+      toImport: Reference<ClassDeclaration>, inComponent: ts.ClassDeclaration,
+      asDirectImport?: boolean): ReadonlyArray<PotentialImport>;
```
Comment (crisbeto) [inline packages/compiler-cli/src/ngtsc/typecheck/api/checker.ts:157]: I'm not a fan of the `asDirectImport` parameter, but I couldn't think of a better way to call it. I need it in the schematic, because I know that some directives that currently aren't standalone will become standalone as a part of the migration so I want the imports to be generated accordingly.

**Comment by dylhunn** [comment]
Thanks, reviewing this today.

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts [hunk 4]
```diff
@@ -598,6 +606,20 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
     return this.typeCheckScopeRegistry.getTypeCheckDirectiveMetadata(new Reference(dir));
   }
 
+  getNgModuleMetadata(module: ts.ClassDeclaration): NgModuleMeta|null {
+    if (!isNamedClassDeclaration(module)) {
+      return null;
+    }
+    return this.metaReader.getNgModuleMetadata(new Reference(module));
```
Comment (dylhunn) [inline packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:613]: Creating a new reference to perform the lookup works fine for locally declared ngModules, but may not work for ngModules coming from external dts file (i.e. anything not in the current compilation unit). This also applies to the other `get*Metadata` methods. I'm not sure if this matters to you -- for more details, look in the dts metadata reader.

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts (hunk 1 above)
Comment (dylhunn) [reply packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:760]: Yes, this does seem like an error. Curious it didn't cause the previous PR to misbehave.

File: packages/compiler-cli/src/ngtsc/typecheck/api/checker.ts (hunk 3 above)
Comment (dylhunn) [reply packages/compiler-cli/src/ngtsc/typecheck/api/checker.ts:157]: I agree it's not the best API, but if you need it then you need it. Can you add a default option of `false` and give it a scarier name so it's not used elsewhere? Maybe something like `forceStandaloneImport`, and then add a `TODO` comment that it can be removed in a standalone world :)

**Review by dylhunn — APPROVED** [review]
LGTM with notes above.

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts (hunk 4 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:613]: It doesn't really matter for my case. FWIW, I was following what we were doing in `getDirectiveMetadata`.

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts [hunk 5]
```diff
@@ -712,18 +734,22 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
   }
 
   getPotentialImportsFor(
-      toImport: PotentialDirective|PotentialPipe,
-      inContext: ts.ClassDeclaration): ReadonlyArray<PotentialImport> {
+      toImport: Reference<ClassDeclaration>, inContext: ts.ClassDeclaration,
+      forceStandaloneImport = false): ReadonlyArray<PotentialImport> {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:738]:
```suggestion
      forceStandaloneImport): ReadonlyArray<PotentialImport> {
```
General style for the compiler is to avoid default values for parameters.

Instead of a boolean parameter, you might consider something like:

```typescript
export enum PotentialImportMode {
  Normal,
  ForceDirect,
}
```

as it carries more semantic information than a boolean flag.

**Review by alxhub — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 6beff5e8d708b575c38320fc8f997bd304a738f3.
